### PR TITLE
refactorings around bytecode representation

### PR DESCRIPTION
**Pull request by mhaupt**
This is a cleanup and refactoring PR, in preparation of more test coverage and functionality around disassembly and compiler behaviour.

### CODE REVIEW

File: src/som/interpreter/Bytecodes.java [hunk 1]
```diff
@@ -24,62 +24,78 @@
...
+  public static final byte POP_LOCAL        = 9;
+  public static final byte POP_ARGUMENT     = 10;
+  public static final byte POP_FIELD        = 11;
+  public static final byte SEND             = 12;
+  public static final byte SUPER_SEND       = 13;
+  public static final byte RETURN_LOCAL     = 14;
+  public static final byte RETURN_NON_LOCAL = 15;
+
+  private static final String[] PADDED_BYTECODE_NAMES = new String[] {
+    "HALT", "DUP", "PUSH_LOCAL", "PUSH_ARGUMENT", "PUSH_FIELD", "PUSH_BLOCK",
```
Comment (smarr) [inline src/som/interpreter/Bytecodes.java]: These don't seem to be padded anymore. Is this on purpose?

**Comment by mhaupt** [comment]
Oh ouch, bad glitch; thanks. That'll require another look. And tests (on the to-do list).

**Comment by mhaupt** [comment]
IntelliJ user here. What exactly is Checkstyle complaining about? The log should be more verbose.

**Comment by mhaupt** [comment]
Running Checkstyle locally:

```
$ ant checkstyle
Buildfile: /Users/haupt/SOMs/mhaupt/som-java/build.xml

checkstyle-jar:
      [get] Getting: http://tenet.dl.sourceforge.net/project/checkstyle/checkstyle/7.8.1/checkstyle-7.8.1-all.jar
      [get] To: /Users/haupt/SOMs/mhaupt/som-java/libs/checkstyle-7.8.1-all.jar

checkstyle:
[checkstyle] Running Checkstyle 7.8.1 on 42 files

BUILD SUCCESSFUL
Total time: 13 seconds
```

**Comment by mhaupt** [comment]
It's actually the eclipseformat command that's failing. How can that be run without Eclipse being present? Does it have to be part of the CI?

**Comment by smarr** [comment]
To make the log more verbose, you could just add a `git diff` in case of failure. That will show the changes made to the files.

You can also use a formatter plugin in Intelij, for instance: https://plugins.jetbrains.com/plugin/6546-eclipse-code-formatter

Or, I could fix the formatting for you.

And, yes, I would very much prefer to have code format checking as part of the CI setup.
That way, I don't need to adjust to different code styles between the different SOM implementations in Java.

**Comment by mhaupt** [comment]
The git diff won't help; of course I need to know which change in particular violates the formatting rules.

I'll take a look at the Eclipse formatter for IntelliJ. 

I'm perfectly fine with style and format checks being part of CI. Doesn't Eclipse formatter have a mode where it enforces formatting rules?

**Comment by mhaupt** [comment]
("Enforces" as in "applies".)

**Comment by smarr** [comment]
That's exactly what `ant eclipseformat` does.

Please see: https://github.com/SOM-st/som-java/blob/master/build.xml#L47

I think, the real problem here is that the check for changes is done with `git status`, and ant doesn't actually show the diff once the status check indicated changes to the files. Perhaps the two operations should be reordered. Not sure though what the effect on a clean build is.

**Comment by mhaupt** [comment]
I'm confused ... if ant eclipseformat applies formatting rules, can't it just do an additional commit with the corrected formatting? I may be misunderstanding something here.

**Comment by smarr** [comment]
You need to do that locally though.

That's what I was trying to say with this line:

> You can run the formatter locally either as part of Eclipse (should be part of the project settings) or manually on the command line (requires `export ECLIPSE_EXE=` to be set).

**Comment by mhaupt** [comment]
Not quite, sorry.

```
$ export ECLIPSE_EXE=
$ ant eclipseformat
Buildfile: /Users/haupt/SOMs/mhaupt/som-java/build.xml

eclipseformat:

BUILD FAILED
/Users/haupt/SOMs/mhaupt/som-java/build.xml:34: no executable specified

Total time: 0 seconds
```

**Comment by smarr** [comment]
Right, you have to provide the path to eclipse as value for `ECLIPSE_EXE`.
So, for example `export ECLIPSE_EXE=/Users/smarr/tmp/eclipe/Eclipse.app/Contents/MacOS/eclipse`.

**Comment by mhaupt** [comment]
I mentioned I don't have Eclipse installed, and I'm not going to install it just to be able to format the sources. The way the project is configured right now makes strong assumptions about what IDE is being used; that should probably be relaxed. Let's work together to find a lightweight solution that allows for enforcing formatting rules without a full Eclipse installation. Meanwhile, I'll work with IntelliJ internal capabilities and the plugin.

And the PR looks as if it could be merged now.

**Comment by smarr** [comment]
I am not sure what you expected.
Yes, you need Eclipse for doing the formatting.
As far as I know there is no standalone tool to do that.

If you don't want to have Eclipse on your disk, the Intellij plugin seems to be the only other way.

Either way, if there is a more lightweight solution that works well in Eclipse and Intellij and doesn't require a completely different code style, we can use that in the CI setup.

And about the 'strong assumptions', well, so far, nobody complained...
My main desire is to keep SOM, TruffleSOM, SOMns, TruffleMate, Grace, etc similar in code style.
So that moving patches around isn't absolutely impossible.
And working with students, automatic formatting is absolutely necessary.

**Comment by mhaupt** [comment]
As mentioned, I completely agree with formatting rules. I also agree with applying the same rules across several code bases. I'm interested in a way that makes no assumptions about IDEs and allows for a similarly lightweight approach to code formatting as is already in place for Checkstyle, where only a JAR is downloaded to make it work. That so far no one has raised this only means that, so far, no one has come around who doesn't want to install a full IDE just to use it in headless mode for code formatting. Well, here I am. :-)

OK to merge?

**Comment by smarr** [comment]
Merging: sure, go ahead.

With respect to lightweight. That's a matter of definition. Sure, it's [just a jar](https://github.com/SOM-st/som-java/blob/master/build.xml#L70).
For Eclipse, it's equally lightweight, it's [just a tar](https://github.com/SOM-st/som-java/blob/master/.travis.yml#L9).

Code wise, they are identical.
The real difference is that the jar is a few KB, while the tar is some 200MB or so...

**Comment by mhaupt** [comment]
Precisely. ;-)

**Pull request merged**